rust/main/chains/hyperlane-starknet/src/provider.rs

```diff
@@ -0,0 +1,156 @@
+
+        Self {
+            domain,
+            rpc_client: Arc::new(provider),
+        }
+    }
+
+    /// Get the RPC client.
+    pub fn rpc_client(&self) -> Arc<AnyProvider> {
+        self.rpc_client.clone()
+    }
+
+    /// Get the hyperlane domain.
+    pub fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+}
+
+impl HyperlaneChain for StarknetProvider {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.clone())
+    }
+}
+
+#[async_trait]
+impl HyperlaneProvider for StarknetProvider {
+    #[instrument(err, skip(self))]
+    async fn get_block_by_height(&self, height: u64) -> ChainResult<BlockInfo> {
+        let block = self
+            .rpc_client()
+            .get_block_with_tx_hashes(BlockId::Number(height))
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        match block {
+            MaybePendingBlockWithTxHashes::Block(b) => Ok(BlockInfo {
+                hash: H256::from_slice(b.block_hash.to_bytes_be().as_slice()),
+                timestamp: b.timestamp,
+                number: b.block_number,
+            }),
+            _ => Err(HyperlaneStarknetError::InvalidBlock.into()),
+        }
+    }
+
+    #[instrument(err, skip(self))]
+    async fn get_txn_by_hash(&self, hash: &H512) -> ChainResult<TxnInfo> {
+        let tx = self
+            .rpc_client()
+            .get_transaction_by_hash(
+                FieldElement::from_byte_slice_be(hash.as_bytes())
+                    .map_err(Into::<HyperlaneStarknetError>::into)?,
+            )
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+
+        let receipt = self
+            .rpc_client()
+            .get_transaction_receipt(tx.transaction_hash())
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+
+        match receipt {
+            MaybePendingTransactionReceipt::Receipt(tx_receipt) => match tx_receipt {
+                TransactionReceipt::Invoke(invoke_receipt) => Ok(TxnInfo {
+                    hash: H512::from_slice(tx.transaction_hash().to_bytes_be().as_slice()),
+                    gas_limit: U256::one(),
+                    max_priority_fee_per_gas: None,
+                    max_fee_per_gas: None,
+                    gas_price: None,
+                    nonce: 0,
+                    sender: H256::zero(),
+                    recipient: None,
+                    raw_input_data: None,
+                    receipt: Some(TxnReceiptInfo {
+                        gas_used: U256::from_big_endian(
+                            invoke_receipt.actual_fee.amount.to_bytes_be().as_slice(),
+                        ),
+                        cumulative_gas_used: U256::zero(),
+                        effective_gas_price: None,
+                    }),
```

Comment:
It will be great to populate all these fields

Reply:
Unfortunately that's all the info available in the receipt on starknet

Reply:
right but it does involve another rpc call

Reply:
that's ok since we'd need this data for the explorer integration 
cc @ameten

Reply:
Unfortunately many fields you referenced are only accesed when building the tx but not through the receipt

Reply:
I've filled most of the fields that were low hanging fruits. 

A few remarks for the missing ones:
- `gas_price` this is a tricky one since the gas price used by the tx is the one defined by the gateway, so its not available on the rpc (yet), we can add it when it is
- We dont have access to the `resource_bounds` field so gas related fields cannott be populated unfortunately, also there is no notion of gas limit really on starknet neither of priority fees
- Last thing is about the recipient, so on Starknet there is native account abstraction meaning that based on the type of account it is we will have to decode what's the contract being called, can be done but a bit dirty and not maintanable, I can create an issue for further investigation